File: internal/lexer/lexer.go
package lexer

import (
	"fmt"
	"strings"
	"unicode"

	"walklang/internal/ast"
)

type TokenKind string

const (
	TokenName   TokenKind = "NAME"
	TokenNumber TokenKind = "NUMBER"
	TokenString TokenKind = "STRING"
	TokenSymbol TokenKind = "SYMBOL"
)

type Token struct {
	Kind     TokenKind
	Value    string
	Location ast.Location
}

type Line struct {
	Indent   int
	Tokens   []Token
	Location ast.Location
}

func Lex(source string, filename string) ([]Line, error) {
	var lines []Line
	for lineNumber, raw := range strings.Split(source, "\n") {
		line := lineNumber + 1
		if tab := strings.IndexRune(raw, '\t'); tab >= 0 {
			return nil, errorAt(ast.Location{Filename: filename, Line: line, Column: tab + 1}, "syntax error: tabs are invalid in v0 indentation")
		}

		body := stripComment(raw)
		if strings.TrimSpace(body) == "" {
			continue
		}

		indent := len(body) - len(strings.TrimLeft(body, " "))
		tokens, err := tokenizeLine(body[indent:], filename, line, indent+1)
		if err != nil {
			return nil, err
		}
		if len(tokens) > 0 {
			lines = append(lines, Line{Indent: indent, Tokens: tokens, Location: tokens[0].Location})
		}
	}
	return lines, nil
}

func stripComment(line string) string {
	inString := false
	escaped := false
	for i, r := range line {
		if inString {
			switch {
			case escaped:
				escaped = false
			case r == '\\':
				escaped = true
			case r == '\'':
				inString = false
			}
			continue
		}
		if r == '\'' {
			inString = true
			continue
		}
		if r == '#' {
			return line[:i]
		}
	}
	return line
}

func tokenizeLine(text string, filename string, line int, startColumn int) ([]Token, error) {
	var tokens []Token
	for i := 0; i < len(text); {
		ch := rune(text[i])
		column := startColumn + i

		if unicode.IsSpace(ch) {
			i++
			continue
		}

		location := ast.Location{Filename: filename, Line: line, Column: column}
		if ch == '\'' {
			value, next, err := readString(text, i, location)
			if err != nil {
				return nil, err
			}
			tokens = append(tokens, Token{Kind: TokenString, Value: value, Location: location})
			i = next
			continue
		}

		if startsNumber(text, i) {
			value, next := readNumber(text, i)
			tokens = append(tokens, Token{Kind: TokenNumber, Value: value, Location: location})
			i = next
			continue
		}

		if i+1 < len(text) {
			two := text[i : i+2]
			if two == ">=" || two == "<=" || two == "==" || two == "!=" {
				tokens = append(tokens, Token{Kind: TokenSymbol, Value: two, Location: location})
				i += 2
				continue
			}
		}

		if strings.ContainsRune("()[],:=+-*/^><?.", ch) {
			tokens = append(tokens, Token{Kind: TokenSymbol, Value: string(ch), Location: location})
			i++
			continue
		}

		if unicode.IsLetter(ch) || ch == '_' {
			value, next := readName(text, i)
			tokens = append(tokens, Token{Kind: TokenName, Value: value, Location: location})
			i = next
			continue
		}

		return nil, errorAt(location, "syntax error: unexpected character %q", ch)
	}
	return tokens, nil
}

func startsNumber(text string, i int) bool {
	if text[i] >= '0' && text[i] <= '9' {
		return true
	}
	return text[i] == '-' && i+1 < len(text) && text[i+1] >= '0' && text[i+1] <= '9' &&
		(i == 0 || unicode.IsSpace(rune(text[i-1])) || strings.ContainsRune("([,=:", rune(text[i-1])))
}

func readNumber(text string, i int) (string, int) {
	start := i
	if text[i] == '-' {
		i++
	}
	seenDot := false
	for i < len(text) {
		if text[i] == '.' && !seenDot {
			seenDot = true
			i++
			continue
		}
		if text[i] < '0' || text[i] > '9' {
			break
		}
		i++
	}
	return text[start:i], i
}

func readName(text string, i int) (string, int) {
	start := i
	for i < len(text) {
		r := rune(text[i])
		if !unicode.IsLetter(r) && !unicode.IsDigit(r) && r != '_' {
			break
		}
		i++
	}
	return text[start:i], i
}

func readString(text string, i int, location ast.Location) (string, int, error) {
	i++
	var out strings.Builder
	escaped := false
	for i < len(text) {
		ch := text[i]
		if escaped {
			switch ch {
			case 'n':
				out.WriteByte('\n')
			case 't':
				out.WriteByte('\t')
			case '\\', '\'':
				out.WriteByte(ch)
			default:
				out.WriteByte(ch)
			}
			escaped = false
		} else if ch == '\\' {
			escaped = true
		} else if ch == '\'' {
			return out.String(), i + 1, nil
		} else {
			out.WriteByte(ch)
		}
		i++
	}
	return "", 0, errorAt(location, "syntax error: unterminated string")
}

func errorAt(location ast.Location, format string, args ...any) error {
	return fmt.Errorf("%s:%d: %s", location.Filename, location.Line, fmt.Sprintf(format, args...))
}

File: compiler/lex/lexer.cpp
#include "lex/lexer.h"

#include <sstream>
#include <string>
#include <string_view>

namespace walk::lex {
namespace {

constexpr const char* kSyntaxDiagnostic = "W1001";

bool is_ascii_space(char ch) {
    return ch == ' ' || ch == '\n' || ch == '\r' || ch == '\t' || ch == '\v' || ch == '\f';
}

bool is_ascii_digit(char ch) {
    return ch >= '0' && ch <= '9';
}

bool is_ascii_letter(char ch) {
    return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z');
}

bool is_name_start(char ch) {
    return is_ascii_letter(ch) || ch == '_';
}

bool is_name_continue(char ch) {
    return is_name_start(ch) || is_ascii_digit(ch);
}

bool is_symbol(char ch) {
    return std::string_view("()[],:=+-*/^><?.").find(ch) != std::string_view::npos;
}

std::string quoted_char(char ch) {
    switch (ch) {
    case '\n':
        return "'\\n'";
    case '\r':
        return "'\\r'";
    case '\t':
        return "'\\t'";
    default:
        return "'" + std::string(1, ch) + "'";
    }
}

void add_error(DiagnosticSet& diagnostics, const SourceFile& source, std::size_t start, std::size_t end, std::string message) {
    diagnostics.add(Diagnostic(DiagnosticSeverity::Error, kSyntaxDiagnostic, std::move(message), source.range_for_offsets(start, end)));
}

std::string strip_comment(std::string_view line) {
    bool in_string = false;
    bool escaped = false;
    for (std::size_t index = 0; index < line.size(); ++index) {
        const char ch = line[index];
        if (in_string) {
            if (escaped) {
                escaped = false;
            } else if (ch == '\\') {
                escaped = true;
            } else if (ch == '\'') {
                in_string = false;
            }
            continue;
        }
        if (ch == '\'') {
            in_string = true;
            continue;
        }
        if (ch == '/' && index + 2 < line.size() && line.substr(index, 3) == "///") {
            return std::string(line.substr(0, index));
        }
        if (ch == '#') {
            return std::string(line.substr(0, index));
        }
    }
    return std::string(line);
}

bool is_blank(std::string_view text) {
    for (const char ch : text) {
        if (!is_ascii_space(ch)) {
            return false;
        }
    }
    return true;
}

bool starts_number(std::string_view text, std::size_t index) {
    if (is_ascii_digit(text[index])) {
        return true;
    }
    if (text[index] != '-' || index + 1 >= text.size() || !is_ascii_digit(text[index + 1])) {
        return false;
    }
    if (index == 0) {
        return true;
    }
    const char previous = text[index - 1];
    return is_ascii_space(previous) || std::string_view("([,=:").find(previous) != std::string_view::npos;
}

std::pair<std::string, std::size_t> read_number(std::string_view text, std::size_t index) {
    const std::size_t start = index;
    if (text[index] == '-') {
        ++index;
    }
    bool seen_dot = false;
    while (index < text.size()) {
        if (text[index] == '.' && !seen_dot) {
            seen_dot = true;
            ++index;
            continue;
        }
        if (!is_ascii_digit(text[index])) {
            break;
        }
        ++index;
    }
    return {std::string(text.substr(start, index - start)), index};
}

std::pair<std::string, std::size_t> read_name(std::string_view text, std::size_t index) {
    const std::size_t start = index;
    while (index < text.size() && is_name_continue(text[index])) {
        ++index;
    }
    return {std::string(text.substr(start, index - start)), index};
}

struct StringRead {
    std::string value;
    std::size_t next = 0;
    bool ok = false;
};

StringRead read_string(std::string_view text, std::size_t index, DiagnosticSet& diagnostics, const SourceFile& source, std::size_t base_offset) {
    const std::size_t quote_index = index;
    ++index;
    std::ostringstream output;
    bool escaped = false;
    int interpolation_depth = 0;
    bool in_interpolation_string = false;
    bool interpolation_escaped = false;

    while (index < text.size()) {
        const char ch = text[index];
        if (interpolation_depth > 0) {
            output << ch;
            if (in_interpolation_string) {
                if (interpolation_escaped) {
                    interpolation_escaped = false;
                } else if (ch == '\\') {
                    interpolation_escaped = true;
                } else if (ch == '\'') {
                    in_interpolation_string = false;
                }
                ++index;
                continue;
            }
            switch (ch) {
            case '\'':
                in_interpolation_string = true;
                break;
            case '{':
                ++interpolation_depth;
                break;
            case '}':
                --interpolation_depth;
                break;
            default:
                break;
            }
        } else if (escaped) {
            switch (ch) {
            case 'n':
                output << '\n';
                break;
            case 't':
                output << '\t';
                break;
            case '\\':
            case '\'':
                output << ch;
                break;
            default:
                output << ch;
                break;
            }
            escaped = false;
        } else if (ch == '\\') {
            escaped = true;
        } else if (ch == '\'') {
            return {output.str(), index + 1, true};
        } else if (ch == '{' && index + 1 < text.size() && text[index + 1] == '{') {
            output << "{{";
            index += 2;
            continue;
        } else if (ch == '{') {
            interpolation_depth = 1;
            output << ch;
        } else {
            output << ch;
        }
        ++index;
    }

    add_error(diagnostics, source, base_offset + quote_index, base_offset + quote_index + 1, "syntax error: unterminated string");
    return {};
}

std::vector<Token> tokenize_line(std::string_view text, const SourceFile& source, std::size_t base_offset, DiagnosticSet& diagnostics) {
    std::vector<Token> tokens;
    for (std::size_t index = 0; index < text.size();) {
        const char ch = text[index];
        if (is_ascii_space(ch)) {
            ++index;
            continue;
        }

        const std::size_t token_start = index;
        if (ch == '\'') {
            const StringRead string = read_string(text, index, diagnostics, source, base_offset);
            if (!string.ok) {
                return {};
            }
            tokens.push_back({TokenKind::String, string.value, source.range_for_offsets(base_offset + token_start, base_offset + string.next)});
            index = string.next;
            continue;
        }

        if (starts_number(text, index)) {
            const auto [value, next] = read_number(text, index);
            tokens.push_back({TokenKind::Number, value, source.range_for_offsets(base_offset + token_start, base_offset + next)});
            index = next;
            continue;
        }

        if (index + 1 < text.size()) {
            const std::string_view two = text.substr(index, 2);
            if (two == ">=" || two == "<=" || two == "==" || two == "!=") {
                tokens.push_back({TokenKind::Symbol, std::string(two), source.range_for_offsets(base_offset + token_start, base_offset + token_start + 2)});
                index += 2;
                continue;
            }
        }

        if (is_symbol(ch)) {
            tokens.push_back({TokenKind::Symbol, std::string(1, ch), source.range_for_offsets(base_offset + token_start, base_offset + token_start + 1)});
            ++index;
            continue;
        }

        if (is_name_start(ch)) {
            const auto [value, next] = read_name(text, index);
            tokens.push_back({TokenKind::Name, value, source.range_for_offsets(base_offset + token_start, base_offset + next)});
            index = next;
            continue;
        }

        add_error(diagnostics, source, base_offset + token_start, base_offset + token_start + 1, "syntax error: unexpected character " + quoted_char(ch));
        return {};
    }
    return tokens;
}

}  // namespace

std::string token_kind_name(TokenKind kind) {
    switch (kind) {
    case TokenKind::Name:
        return "NAME";
    case TokenKind::Number:
        return "NUMBER";
    case TokenKind::String:
        return "STRING";
    case TokenKind::Symbol:
        return "SYMBOL";
    }
    return "UNKNOWN";
}

LexResult lex_source(const SourceFile& source) {
    LexResult result;
    const std::string text(source.text());

    std::size_t line_start = 0;
    while (line_start <= text.size()) {
        std::size_t line_end = text.find('\n', line_start);
        if (line_end == std::string::npos) {
            line_end = text.size();
        }
        const std::string_view raw(text.data() + line_start, line_end - line_start);

        const std::size_t tab = raw.find('\t');
        if (tab != std::string_view::npos) {
            add_error(result.diagnostics, source, line_start + tab, line_start + tab + 1, "syntax error: tabs are invalid in indentation");
            return result;
        }

        const std::string body = strip_comment(raw);
        if (!is_blank(body)) {
            std::size_t indent = 0;
            while (indent < body.size() && body[indent] == ' ') {
                ++indent;
            }
            std::vector<Token> tokens = tokenize_line(std::string_view(body).substr(indent), source, line_start + indent, result.diagnostics);
            if (result.diagnostics.has_errors()) {
                return result;
            }
            if (!tokens.empty()) {
                result.lines.push_back({indent, std::move(tokens), source.range_for_offsets(line_start + indent, line_start + indent + 1)});
            }
        }

        if (line_end == text.size()) {
            break;
        }
        line_start = line_end + 1;
    }

    return result;
}

}  // namespace walk::lex
